### PR TITLE
qos proposal header files

### PR DESCRIPTION
**Pull request by lguohan**
squash multiple commits from ashok into one

### CODE REVIEW

File: sai/inc/saiswitch.h [hunk 1]
```diff
@@ -176,6 +176,17 @@ typedef enum _sai_switch_attr_t
     /* Maximum number of LAGs that can be created per switch [uint32_t] */
     SAI_SWITCH_ATTR_MAX_LAG_NUMBER,
 
+    /** Maximum traffic classes limit*/
+    SAI_SWITCH_ATTR_QOS_MAX_NUMBER_OF_TRAFFIC_CLASSES,
+
+    /** HQOS - Maximum Number of Hierarchy scheduler 
+     *  group levels(depth) supported.*/
```
Comment (lguohan) [inline sai/inc/saiswitch.h]: type should be [sai_uint32_t]?

File: sai/inc/saiswitch.h [hunk 2]
```diff
@@ -176,6 +176,17 @@ typedef enum _sai_switch_attr_t
...
+    /** Maximum traffic classes limit*/
+    SAI_SWITCH_ATTR_QOS_MAX_NUMBER_OF_TRAFFIC_CLASSES,
+
+    /** HQOS - Maximum Number of Hierarchy scheduler 
+     *  group levels(depth) supported.*/ 
+    SAI_SWITCH_ATTR_QOS_MAX_NUMBER_OF_SCHEDULER_GROUP_HIERARCHY_LEVELS,
+
+    /** HQOS - Maximum number of scheduler groups supported on 
+     * each Hierarchy level [sai_scheduler_group_count_t] */  
+    SAI_SWITCH_ATTR_QOS_MAX_NUMBER_OF_SCHEDULER_GROUPS_PER_HIERARCHY_LEVEL,
```
Comment (lguohan) [inline sai/inc/saiswitch.h]: similiar to port list, can we make this as sai_u32_list_t to return number of scheduler groups for all hierarchy levels in one shoot. Then, we can remove the struct sai_scheduler_group_count_t?

File: sai/inc/saitypes.h [hunk 3]
```diff
@@ -325,6 +331,67 @@ typedef struct _sai_port_breakout_t
...
+
+    /* Dot1p value */
+    sai_uint8_t dot1p;
+
+    /* Color of the packet */
+    sai_packet_color_t color;
+
+    /* Egress port queue UOID is not known at the time of map creation. 
+     * Using queue index for maps.*/
+    sai_cos_t    queue_index;
```
Comment (lguohan) [inline sai/inc/saitypes.h]: sai_cos_t => sai_queue_id_t?

Comment (lguohan) [inline sai/inc/saitypes.h]: sai_cos_t => sai_queue_index_t?

File: sai/inc/saischedulergroup.h [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    SAI_SCHEDULER_GROUP_ATTR_CHILD_LIST = 0x00000002,
+
+    /** READ-WRITE */
+
+    /** Scheduler group on port [sai_object_id_t]
+       MANDATORY_ON_CREATE,  CREATE_ONLY */
+    SAI_SCHEDULER_GROUP_ATTR_PORT_ID = 0x00000003,
+
+    /** Scheduler group level
+       MANDATORY_ON_CREATE,  CREATE_ONLY */
```
Comment (lguohan) [inline sai/inc/saischedulergroup.h]: type for this attribute?

File: sai/inc/saischeduler.h [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    /* Deficit Weighted Round-Robin Scheduling */
+    SAI_SCHEDULING_DWRR = 0x00000002,
+
+} sai_scheduling_type_t;
+
+/**
+ * @brief Enum defining scheduler attributes.
+ */
+typedef enum _sai_scheduler_attr_t
+{
```
Comment (lguohan) [inline sai/inc/saischeduler.h]: /* READ-WRITE */

File: sai/inc/saischeduler.h [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    /** Scheduling algorithm [sai_scheduling_type_t ], Default WRR*/
+    SAI_SCHEDULER_ATTR_SCHEDULING_ALGORITHM = 0x00000000,
+
+    /** [uint8_t] scheduling algorithm weight, Range [1 - 100]
+        Valid SAI_SCHEDULER_ATTR_SCHEDULING_ALGORITHM = SAI_SCHEDULING_DWRR, 
+        Default Weight = 1 */
+    SAI_SCHEDULER_ATTR_SCHEDULING_WEIGHT = 0x00000001,
+
+    /* sai_meter_type_t, Default bytes/sec */
```
Comment (lguohan) [inline sai/inc/saischeduler.h]: [sai_meter_type_t]

File: sai/inc/saischeduler.h [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    SAI_SCHEDULER_ATTR_SHAPER_TYPE = 0x00000002,
+
+    /** [uint64_t] Guaranteed Bandwidth shape rate [bytes/sec or PPS]
+     0 to disable the values */
+    SAI_SCHEDULER_ATTR_MIN_BANDWIDTH_RATE = 0x00000003,
```
Comment (lguohan) [inline sai/inc/saischeduler.h]: default value for this attribute?

File: sai/inc/saischeduler.h [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+    /** [uint64_t] Guaranteed Burst for Bandwidth shape rate [Bytes or Packets]
+     */
+    SAI_SCHEDULER_ATTR_MIN_BANDWIDTH_BURST_RATE = 0x00000004,
```
Comment (lguohan) [inline sai/inc/saischeduler.h]: default value for this attribute?

File: sai/inc/saitypes.h [hunk 9]
```diff
@@ -325,6 +331,67 @@ typedef struct _sai_port_breakout_t
...
+ * Examples:
+ * dot1p/dscp --> TC
+ * dot1p/dscp --> Color
+ * dot1p/dscp --> TC + Color
+ * Tc --> dot1p/Dscp.
+ * Tc + color --> dot1p/Dscp.
+ * Tc --> Egress Queue.
+ */
+
+typedef struct _sai_qos_map_params_t
```
Comment (lguohan) [inline sai/inc/saitypes.h]: can this be a union?

File: sai/inc/saiswitch.h (hunk 1 above)
Comment (AshokDaparthi) [inline sai/inc/saiswitch.h]: Added

File: sai/inc/saiswitch.h (hunk 2 above)
Comment (AshokDaparthi) [inline sai/inc/saiswitch.h]: Modified

File: sai/inc/saitypes.h (hunk 3 above)
Comment (AshokDaparthi) [inline sai/inc/saitypes.h]: Added type and changed

File: sai/inc/saischedulergroup.h (hunk 4 above)
Comment (AshokDaparthi) [inline sai/inc/saischedulergroup.h]: sai_uint8_t Added

File: sai/inc/saischeduler.h (hunk 5 above)
Comment (AshokDaparthi) [inline sai/inc/saischeduler.h]: Added

File: sai/inc/saischeduler.h (hunk 6 above)
Comment (AshokDaparthi) [inline sai/inc/saischeduler.h]: Done

File: sai/inc/saischeduler.h (hunk 7 above)
Comment (AshokDaparthi) [inline sai/inc/saischeduler.h]: Default is 0, Disable state

File: sai/inc/saitypes.h (hunk 9 above)
Comment (AshokDaparthi) [inline sai/inc/saitypes.h]: Modified the order to solve padding issues.

File: sai/inc/saischeduler.h (hunk 7 above)
Comment (itaibaz) [inline sai/inc/saischeduler.h]: Hi Ashok,
Re :

    /** [uint64_t] Guaranteed Burst for Bandwidth shape rate [Bytes or Packets]
     */
    SAI_SCHEDULER_ATTR_MIN_BANDWIDTH_BURST_RATE = 0x00000004,

    /** [uint64_t] Maximum Burst for Bandwidth shape rate [bytes or Packets]
     */
    SAI_SCHEDULER_ATTR_MAX_BANDWIDTH_BURST_RATE = 0x00000006,

Can you explain the meaning of these 2 attributes ?
Can you please give an example of using them ?

Thanks

**Pull request merged**